File: frontend/src/app/services/map-variable.service.ts
import {Injectable, inject} from '@angular/core';
import {HttpClient} from '@angular/common/http';
import {Observable} from 'rxjs';
import {MapVariable, PicklistValue, CellVariableValue} from '../models/map-variable.model';

@Injectable({providedIn: 'root'})
export class MapVariableService {
  private http = inject(HttpClient);

  private baseUrl(mapId: number): string {
    return `/api/maps/${mapId}/variables`;
  }

  getVariables(mapId: number): Observable<MapVariable[]> {
    return this.http.get<MapVariable[]>(this.baseUrl(mapId));
  }

  createVariable(mapId: number, variable: Partial<MapVariable>): Observable<MapVariable> {
    return this.http.post<MapVariable>(this.baseUrl(mapId), variable);
  }

  updateVariable(mapId: number, variableId: string, patch: Partial<MapVariable>): Observable<MapVariable> {
    return this.http.put<MapVariable>(`${this.baseUrl(mapId)}/${variableId}`, patch);
  }

  deleteVariable(mapId: number, variableId: string): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl(mapId)}/${variableId}`);
  }

  getPicklistValues(mapId: number, variableId: string): Observable<PicklistValue[]> {
    return this.http.get<PicklistValue[]>(`${this.baseUrl(mapId)}/${variableId}/picklist-values`);
  }

  addPicklistValue(mapId: number, variableId: string, label: string): Observable<PicklistValue> {
    return this.http.post<PicklistValue>(`${this.baseUrl(mapId)}/${variableId}/picklist-values`, {label});
  }

  updatePicklistValue(mapId: number, variableId: string, picklistValueId: string, label: string): Observable<PicklistValue> {
    return this.http.put<PicklistValue>(`${this.baseUrl(mapId)}/${variableId}/picklist-values/${picklistValueId}`, {label});
  }

  deletePicklistValue(mapId: number, variableId: string, picklistValueId: string): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl(mapId)}/${variableId}/picklist-values/${picklistValueId}`);
  }
}

File: backend/src/main/java/org/dungeonmaps/DungeonMapster/service/MapVariableService.java
package org.dungeonmaps.DungeonMapster.service;

import org.dungeonmaps.DungeonMapster.model.MapVariable;
import org.dungeonmaps.DungeonMapster.model.PicklistValue;
import org.dungeonmaps.DungeonMapster.repository.CellVariableValueRepository;
import org.dungeonmaps.DungeonMapster.repository.MapVariableRepository;
import org.dungeonmaps.DungeonMapster.repository.PicklistValueRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.Optional;

@Service
public class MapVariableService {

    private static final String[] COLOR_PALETTE = {
            "#e74c3c", "#3498db", "#2ecc71", "#9b59b6", "#f39c12",
            "#1abc9c", "#e91e63", "#ff5722", "#607d8b", "#795548",
            "#00bcd4", "#8bc34a", "#ff9800", "#673ab7", "#009688"
    };

    private final MapVariableRepository variableRepository;
    private final PicklistValueRepository picklistValueRepository;
    private final CellVariableValueRepository cellVariableValueRepository;

    public MapVariableService(MapVariableRepository variableRepository,
                              PicklistValueRepository picklistValueRepository,
                              CellVariableValueRepository cellVariableValueRepository) {
        this.variableRepository = variableRepository;
        this.picklistValueRepository = picklistValueRepository;
        this.cellVariableValueRepository = cellVariableValueRepository;
    }

    public List<MapVariable> getVariablesForMap(Long mapId) {
        return variableRepository.findByMapIdOrderBySortOrder(mapId);
    }

    public Optional<MapVariable> getById(String id) {
        return variableRepository.findById(id);
    }

    @Transactional
    public MapVariable createVariable(Long mapId, MapVariable variable) {
        int count = variableRepository.countByMapId(mapId);
        variable.setMapId(mapId);
        variable.setSortOrder(count);
        if (!"PICKLIST".equals(variable.getDataType())) {
            variable.setShowColorOnCells(false);
        }
        return variableRepository.save(variable);
    }

    @Transactional
    public Optional<MapVariable> updateVariable(String id, MapVariable patch) {
        return variableRepository.findById(id).map(existing -> {
            existing.setName(patch.getName());
            existing.setVisibility(patch.getVisibility());
            existing.setDisplayFormat(patch.getDisplayFormat());
            if ("PICKLIST".equals(existing.getDataType())) {
                existing.setShowColorOnCells(patch.isShowColorOnCells());
            }
            return variableRepository.save(existing);
        });
    }

    @Transactional
    public boolean deleteVariable(String id) {
        if (!variableRepository.existsById(id)) return false;
        cellVariableValueRepository.deleteByVariableId(id);
        picklistValueRepository.deleteByVariableId(id);
        variableRepository.deleteById(id);
        return true;
    }

    public List<PicklistValue> getPicklistValues(String variableId) {
        return picklistValueRepository.findByVariableIdOrderBySortOrder(variableId);
    }

    @Transactional
    public PicklistValue addPicklistValue(String variableId, String label) {
        List<PicklistValue> existing = picklistValueRepository.findByVariableIdOrderBySortOrder(variableId);
        String color = assignColor(existing.size());

        PicklistValue pv = new PicklistValue();
        pv.setVariableId(variableId);
        pv.setLabel(label);
        pv.setColor(color);
        pv.setSortOrder(existing.size());
        return picklistValueRepository.save(pv);
    }

    @Transactional
    public Optional<PicklistValue> updatePicklistValue(String picklistValueId, String label) {
        return picklistValueRepository.findById(picklistValueId).map(pv -> {
            pv.setLabel(label);
            return picklistValueRepository.save(pv);
        });
    }

    @Transactional
    public boolean deletePicklistValue(String picklistValueId) {
        return picklistValueRepository.findById(picklistValueId).map(pv -> {
            cellVariableValueRepository.deleteByVariableIdAndValue(pv.getVariableId(), picklistValueId);
            picklistValueRepository.deleteById(picklistValueId);
            return true;
        }).orElse(false);
    }

    private String assignColor(int index) {
        if (index < COLOR_PALETTE.length) {
            return COLOR_PALETTE[index];
        }
        int r = (int) (Math.random() * 200 + 30);
        int g = (int) (Math.random() * 200 + 30);
        int b = (int) (Math.random() * 200 + 30);
        return String.format("#%02x%02x%02x", r, g, b);
    }
}